### ISSUE TITLE
Improve prioritization so that cores don't spend idle time needlessly

### ISSUE THREAD

**Issue by geoffrey4444** [sxs-collaboration/spectre#4766]
Some cores are idle waiting for messages so they can do work, and this adds up so that cores spend roughly 50% of their time idle, even when load balanced using a weighted z curve, so the total work done per core is about the same on all cores. This is a prioritization problem.

Core B has a whole bunch of tasks. Core A depends on a subset of those tasks. But the tasks A is waiting on are not prioritized, so it can happen that core A has to wait, even though A and B have the same total amount of work.

Tasks are currently prioritized FIFO. Fixing this issue is a followup to #4580 , which improves charm++ messaging to allow us to prioritize tasks: next node first, then next core, then local.

**Comment by nilsvu** [comment]
My understanding is that prioritizing messages to elements on other cores/nodes is not enough, but we have to prioritize all messages sent to elements that have neighbors on other cores/nodes. This is because the element that holds up computation on another core can only proceed once it has received data from _all_ of its neighbors, including same-core neighbors. Maybe this is obvious, but I wanted to make sure it's written down here.

**Comment by geoffrey4444** [comment]
@nilsdeppe did try a quick hack of deprioritizing local elements, suggests minimal (a few %) improvement in idle time reduction. The biggest improvement was in the wave zone elements (lots of elements with work to do).

Regarding @nilsvu remarks above: prioritizing all (internal & external) elements that must send to elements with neighbors on other cores could be done, but you'd have to have each element say whether it has neighbors on another core.

- [ ] Idea: if you have enough elements/core, you can work on some elements while communication is going on. @macedo22 starting to look into reorder within the action for GH RHS computation to i) compute data needing to be sent, ii) send, iii) finish computing RHS. Local time stepping will add some complexities. This might help the few-element cores, by keeping more work for the core to do after sending.

Note: inline entry methods avoid PUP, etc...everything local is synchronous. Original z curve and inline entry methods worked together to minimize communication overhead costs...which dominates for uniform problems. But BBH more dominated by computation costs, thanks to very different step sizes, resolutions in elements

- [x] Idea: disable inline entry methods? set the limit to zero to disable this... @macedo22 will do a run like this (@nilsdeppe will ask)

- [ ] Idea: Kyle wants to choose at runtime round-robin, z-curve + uniform weighting, z-curve + other weighting

Could you diagnose why so much idle time? @nilsdeppe diagnosed, found that cores are waiting for others to send it data. So we need to figure out how to get the data sent earlier to the node that's waiting on it.

Note: we have to try all these things in combination...how can we test these in the least amount of effort?

**Comment by nilsdeppe** [comment]
Disabling `inline` did not alter prioritization. This makes sense since we reschedule/return from the action ourselves rather than relying on Charm++ priorities, so given the implementation there's no difference in _how_ the action is called.

**Comment by geoffrey4444** [comment]
@nilsdeppe did a hack on a branch, and this gave a 3% speedup, wasn't enough that this is worth pushing all the way through local time stepping. This will not be a silver bullet for speedup. Move to low priority.
